analyzer/analyzer.go

```diff
@@ -30,310 +23,131 @@ var (
+	// this struct is accessed by multiple goroutines
+	r.m.Lock()
+	defer r.m.Unlock()
 
-	var s1 v1.ConfigFile
-	if err := json.Unmarshal(configBlob, &s1); err != nil {
-		return xerrors.Errorf("json marshal error: %w", err)
+	if new.OS != nil {
+		r.OS = new.OS
 	}
 
-	info := types.ArtifactInfo{
-		SchemaVersion:   types.ArtifactJSONSchemaVersion,
-		Architecture:    s1.Architecture,
-		Created:         s1.Created.Time,
-		DockerVersion:   s1.DockerVersion,
-		OS:              s1.OS,
-		HistoryPackages: pkgs,
+	if len(new.PackageInfos) > 0 {
+		r.PackageInfos = append(r.PackageInfos, new.PackageInfos...)
 	}
 
-	if err := ac.Cache.PutArtifact(imageID, info); err != nil {
-		return xerrors.Errorf("failed to put image info into the cache: %w", err)
+	if len(new.Applications) > 0 {
+		r.Applications = append(r.Applications, new.Applications...)
 	}
-
-	return nil
 }
 
-type Applier struct {
-	cache cache.LocalArtifactCache
-}
-
-func NewApplier(c cache.LocalArtifactCache) Applier {
-	return Applier{cache: c}
-}
-
-func (a Applier) ApplyLayers(imageID string, diffIDs []string) (types.ArtifactDetail, error) {
-	var layers []types.BlobInfo
-	for _, diffID := range diffIDs {
-		layer, _ := a.cache.GetBlob(diffID)
-		if layer.SchemaVersion == 0 {
-			return types.ArtifactDetail{}, xerrors.Errorf("layer cache missing: %s", diffID)
+func AnalyzeFile(filePath string, info os.FileInfo, opener Opener) (AnalysisResult, error) {
+	var result AnalysisResult
+	for _, analyzer := range analyzers {
+		// filepath extracted from tar file doesn't have the prefix "/"
+		if !analyzer.Required(strings.TrimLeft(filePath, "/"), info) {
+			continue
+		}
+		b, err := opener()
+		if err != nil {
+			return AnalysisResult{}, xerrors.Errorf("unable to open a file (%s): %w", filePath, err)
 		}
-		layers = append(layers, layer)
-	}
-
-	mergedLayer := docker.ApplyLayers(layers)
-	if mergedLayer.OS == nil {
-		return mergedLayer, ErrUnknownOS // send back package and apps info regardless
-	} else if mergedLayer.Packages == nil {
-		return mergedLayer, ErrNoPkgsDetected // send back package and apps info regardless
-	}
-
-	imageInfo, _ := a.cache.GetArtifact(imageID)
-	mergedLayer.HistoryPackages = imageInfo.HistoryPackages
-
-	return mergedLayer, nil
-}
 
-func GetOS(filesMap extractor.FileMap) *types.OS {
-	for _, analyzer := range osAnalyzers {
-		os, err := analyzer.Analyze(filesMap)
+		ret, err := analyzer.Analyze(b)
 		if err != nil {
 			continue
 		}
-		return &os
+		result.Merge(ret.ConvertToResult(analyzer.Name(), filePath))
 	}
-	return nil
+	return result, nil
 }
 
-func GetPackages(filesMap extractor.FileMap) ([]types.PackageInfo, error) {
-	var results []types.PackageInfo
-	for _, analyzer := range pkgAnalyzers {
-		pkgMap, err := analyzer.Analyze(filesMap)
-
-		// Differentiate between a package manager not being found and another error
-		if err != nil && err == ErrNoPkgsDetected {
+func AnalyzeConfig(targetOS types.OS, configBlob []byte) []types.Package {
```

Comment:
Could you remind me again what is this used for?

Reply:
On a sidenote, my IDE says only Alpine currently implements `analyzer.Analyze` what about other OSes?

Reply:
`Analyzer` analyzers files in a container image such as `/var/lib/dpkg/status`, but an image has metadata in `config.json`. This is managed separately from the layer content and includes history such as `RUN`, `COPY` and `ENV` used in Dockerfile. `AnalyzeConfig` is intended to analyze such a history.

Especially, the target is the following case.

```
RUN apk add openssl && \
        wget https://example.com/anysoftware.tar.gz && \
        tar xvf anysoftware.tar.gz && cd anysoftware && \
       ./configure --with-ssl && make && make install && \
       apk del openssl
```

In this case, `openssl` is linked to `anysoftware` statically, but it doesn't exist in the merged layer since this package is deleted. So, `AnalyzeConfig` parses `RUN` lines and extracts deleted packages.

>On a sidenote, my IDE says only Alpine currently implements analyzer.Analyze what about other OSes?
Right. As you can see, `RUN` doesn't show which version is installed. But Alpine stores only the latest version in their repository. So, we can know the version from `created_at`. I think it is difficult for other OSes.